### DISCUSSION TITLE
Scheduled job doesn't Run on iOS (Fine on Android)

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#888] [Q&A] [answered]
Shiny 2.4.0.2960

The JobManager will happi;y run an adhoc job on iOS and Android using IJobManager.RunTask

If I try and schedule an IJob in Android it works just fine however on iOS it does not seem to Run. The constructor gets called and I have an ILogger.LogDebug statement in there and that appears in AppCenter.

None of the logging or notifcations get fired in the Run method.

```
 public async Task Run(JobInfo jobInfo, CancellationToken cancelToken)
        {
            logger.LogDebug("In Run Mehtod");

            await this.notificationManager.Send(
               "Job Started",
               $"{jobInfo.Identifier} Started"
            );

            await this.notificationManager.Send(
                "Job Finished",
                $"{jobInfo.Identifier} Finished"
            );
        }
```

The info.plist has the necessary UIBackgroundModes and BGTaskSchedulerPermittedIdentifiers

```
<key>UIBackgroundModes</key>
<array>
    <string>bluetooth-central</string>
    <string>remote-notification</string>
    <string>fetch</string>
    <string>processing</string>
</array>
<key>BGTaskSchedulerPermittedIdentifiers</key>
<array>
    <string>com.shiny.job</string>
    <string>com.shiny.jobpower</string>
    <string>com.shiny.jobnet</string>
    <string>com.shiny.jobpowernet</string>
  </array>

```

I schedule the job like this, 

```
var job = new JobInfo(typeof(Jobs.SyncJob), "SyncJob");
job.BatteryNotLow = false;
job.DeviceCharging = false;
job.RequiredInternetAccess = InternetAccess.Any;
job.RunOnForeground = false;
await jobManager.Register(job);
```

As well as some Push Notifications methods mapped in AppDelegate I also have 

```
public override void PerformFetch(UIApplication application, Action<UIBackgroundFetchResult> completionHandler) => this.ShinyPerformFetch(completionHandler);

public override void HandleEventsForBackgroundUrl(UIApplication application, string sessionIdentifier, Action completionHandler) => this.ShinyHandleEventsForBackgroundUrl(sessionIdentifier, completionHandler);
```

Is there anything else I might be missing on iOS

**Comment by aritchie** [comment, marked as answer]
So there's a couple of catches to iOS - the first is that "scheduled" is a bad term to use.  Periodic is the correct term to use.  iOS determines how often your user actually uses your app and then decides when things can/should run.  

If you are using a mac, try running in the iOS simulator.  Once you are debugging in the simulator, go to visual studio for mac, under debug -> simulator iOS background fetch.  If they run here, all I can tell you is that iOS will run this but not with any sort of guarantees for time.  

A couple of other notes here:
* You also can't assume that AppCenter will push the logged data immediately.  It may not have time in the scope of the job running to do so.
* You don't need HandleEventsForBackgroundUrl - that is http background downloads & uploads
* PerformFetch is a fallback for older platform but not as necessary anymore

**Reply by aritchie** [reply]
Also for notifications, especially on iOS, make sure you've run notificationManager.RequestAccess in the UI otherwise your notifications won't run

**Reply by munkii** [reply]
"_So there's a couple of catches to iOS - the first is that "scheduled" is a bad term to use. Periodic is the correct term to use. iOS determines how often your user actually uses your app and then decides when things can/should run._"

It was this. 

I knew that it was when iOS decided but I had no idea it would be so long. I'd walked away from my desk and driven to the vet. About 45 mins after I had expected it to fire. It fired!

Thanks for all the extra info

**Reply by aritchie** [reply]
It won't necessarily be that long.  They use A LOT of metrics in determining when it runs.  You can have jobs running every 3-5mins if your app is used enough.
